daemon/algod/api/server/v2/handlers.go

```diff
@@ -1114,6 +1115,70 @@ func (v2 *Handlers) GetApplicationByID(ctx echo.Context, applicationID uint64) e
 	return ctx.JSON(http.StatusOK, response)
 }
 
+// GetApplicationBoxes returns the box names of an application
+// (GET /v2/applications/{application-id}/boxes)
+func (v2 *Handlers) GetApplicationBoxes(ctx echo.Context, applicationID uint64, params generated.GetApplicationBoxesParams) error {
+	appIdx := basics.AppIndex(applicationID)
+	ledger := v2.Node.LedgerForAPI()
+	lastRound := ledger.Latest()
+	keyPrefix := logic.MakeBoxKey(appIdx, "")
+
+	maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication
+
+	isNilOrZero := func(a *uint64) bool { return a == nil || *a == 0 }
+	nilToZero := func(a *uint64) uint64 {
+		if a == nil {
+			return 0
+		}
+		return *a
+	}
+
+	// compute `maxReturnNum` from `param.Max` and `maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication`
+	// `maxReturnNum == 0` for returning everything
+	// - `param.Max == nil || *param.Max == 0` and `maxBoxThreshold == 0`,
+	//       return everything
+	// - `param.Max == nil || *param.Max == 0` and `maxBoxThreshold > 0`,
+	//       returns up to `maxBoxThreshold`, check if exceeds
+	// - `*param.Max > 0` and `maxBoxThreshold == 0`,
+	//       return up to `*param.Max`
+	// - `*param.Max > 0` and `maxBoxThreshold > 0`,
+	//       return up to min(*param.Max, maxBoxThreshold),
+	//       if return size is `maxBoxThreshold`, check if exceeds
+
+	var boxKeys []string
+	var err error
+
+	followsParams := (isNilOrZero(params.Max) && maxBoxThreshold == 0) ||
+		(!isNilOrZero(params.Max) && (maxBoxThreshold > *params.Max || maxBoxThreshold == 0))
+
+	if followsParams {
+		boxKeys, err = ledger.LookupKeysByPrefix(lastRound, keyPrefix, nilToZero(params.Max))
+	} else {
+		boxKeys, err = ledger.LookupKeysByPrefix(lastRound, keyPrefix, maxBoxThreshold+1)
+		if uint64(len(boxKeys)) > maxBoxThreshold {
+			v2.Log.Info("MaxAPIBoxPerApplication limit %d exceeded", maxBoxThreshold)
+			return ctx.JSON(http.StatusBadRequest, generated.ErrorResponse{
+				Message: "Result limit exceeded",
+				Data:    nil,
+			})
+		}
+	}
```

Comment:
From previous discussion with @jannotti and @algochoi , we would like to follow the previous convention:
If we are querying for some resource that is exceeding Algod config, then fail early.

This code first sees if the amount of box it is looking for is dominated by Algod config, or query parameter.
- If it is dominated by Algod config, we additionally ask for `1 + maxBoxThreshold` (which is the config).
  We can tell from the size of box names returned, to see if limit is exceeds.
- Otherwise, we just get amount of box names from query parameter, with no additional check.

Let me know if this implementation follows the existing implementation design.